src/legacy/core_plugins/kibana/public/discover/np_ready/components/discover_grid/discover_grid.tsx

```diff
@@ -0,0 +1,478 @@
+  sort,
+  indexPattern,
+  ariaLabelledBy,
+  searchTitle,
+  searchDescription,
+  useShortDots,
+  onSort,
+  sampleSize,
+  onFilter,
+  getContextAppHref,
+  onRemoveColumn,
+  onAddColumn,
+  showTimeCol,
+}: Props) {
+  const lowestPageSize = 50;
+  const timeString = i18n.translate('kbn.discover.timeLabel', {
+    defaultMessage: 'Time',
+  });
+  const [flyoutRow, setFlyoutRow] = useState<ElasticSearchHit | undefined>(undefined);
+
+  const dataGridColumns = columns.map(
+    (columnName, i): EuiDataGridColumn => {
+      const column: EuiDataGridColumn = {
+        id: columnName,
+        schema: indexPattern.getFieldByName(columnName)?.type,
+      };
+
+      // Default DataGrid schemas: boolean, numeric, datetime, json, currency
+      // Default indexPattern types: KBN_FIELD_TYPES in src/plugins/data/common/kbn_field_types/types.ts
+      switch (column.schema) {
+        case 'date':
+          column.schema = 'datetime';
+          break;
+        case 'number':
+          column.schema = 'numeric';
+          break;
+        case '_source':
+        case 'object':
+          column.schema = kibanaJSON;
+          break;
+        case 'geo_point':
+          column.schema = geoPoint;
+          break;
+        default:
+          column.schema = undefined;
+          break;
+      }
+
+      if (useShortDots) {
+        column.display = <>{shortenDottedString(columnName)}</>;
+      }
+
+      return column;
+    }
+  );
+
+  // Discover always injects a Time column as the first item (unless advance settings turned it off)
+  // Have to guard against this to allow users to request the same column again later
+  if (showTimeCol) {
+    dataGridColumns.unshift({ id: timeString, schema: 'datetime', initialWidth: 172 });
+  }
+
+  /**
+   * Pagination
+   */
+  const [pagination, setPagination] = useState({ pageIndex: 0, pageSize: lowestPageSize });
+  const onChangeItemsPerPage = useCallback(
+    pageSize => setPagination(paginationData => ({ ...paginationData, pageSize })),
+    [setPagination]
+  );
+  const onChangePage = useCallback(
+    pageIndex => setPagination(paginationData => ({ ...paginationData, pageIndex })),
+    [setPagination]
+  );
+
+  /**
+   * Sorting
+   */
+  const sortingColumns = useMemo(() => sort.map(([id, direction]) => ({ id, direction })), [sort]);
+  const onTableSort = useCallback(
+    sortingColumnsData => {
+      onSort(sortingColumnsData.map(({ id, direction }: SortObj) => [id, direction]));
+    },
+    [onSort]
+  );
+
+  /**
+   * Visibility and order
+   */
+  const [visibleColumns, setVisibleColumns] = useState(dataGridColumns.map(obj => obj.id));
+  const mounted = React.useRef(false);
+  useEffect(() => {
+    // every time a column is added, make it visible
+    if (!mounted.current) {
+      mounted.current = true;
+    } else {
+      setVisibleColumns(dataGridColumns.map(obj => obj.id));
+    }
+  }, [dataGridColumns.length]); // eslint-disable-line react-hooks/exhaustive-deps
```

Comment:
So, when adding the first column, there is no change in length of columns, it switches between `['_source']` and `['myLovelyFirstColumn']`, so the visible columns aren't set. The following code should fix this (using `isEqual` of lodash)

```TypeScript
 useEffect(() => {
    // every time a column is added, make it visible
    if (!mounted.current) {
      mounted.current = true;
    } else {
      const newColumns = dataGridColumns.map(obj => obj.id);
      if (!isEqual(newColumns, visibleColumns)) {
        setVisibleColumns(newColumns);
      }
    }
  }, [dataGridColumns, visibleColumns]);
```